### PR TITLE
Enhanced bytes limiter with data type param

### PR DESCRIPTION
**Pull request by justinjung04**
* [ ] I added CHANGELOG entry for this change.
* [x] Change is not relevant to the end user.

## Changes

Introducing two main changes to enhance bytes-based limiting:
1. Changed `BytesLimiter.Reserve(num uint64) err` to `BytesLimiter.ReserveWithType(num uint64, dataType storeDataType) err` such that the bytes limiter logic to be more flexible. This is useful given that processing of postings, series and chunks have different impact to pod resource usage. For example, user can have a bytes limiter that has different weights for different data types.
2. nit: Created `queryStats.add` func to update query stats info more consistently

## Verification

Added unit test and e2e test.

### CODE REVIEW

File: pkg/store/bucket.go [hunk 1]
```diff
@@ -3534,7 +3543,7 @@ func (r *bucketChunkReader) load(ctx context.Context, res []seriesEntry, aggrs [
 
 // loadChunks will read range [start, end] from the segment file with sequence number seq.
 // This data range covers chunks starting at supplied offsets.
-func (r *bucketChunkReader) loadChunks(ctx context.Context, res []seriesEntry, aggrs []storepb.Aggr, seq int, part Part, pIdxs []loadIdx, calculateChunkChecksum bool, bytesLimiter BytesLimiter, tenant string) error {
+func (r *bucketChunkReader) loadChunks(ctx context.Context, res []seriesEntry, aggrs []storepb.Aggr, seq int, part Part, pIdxs []loadIdx, calculateChunkChecksum bool, bytesLimiters []BytesLimiter, tenant string) error {
```
Comment (harry671003) [inline pkg/store/bucket.go:3546]: Instead of passing []BytesLimiter here, we could still pass an implemetation of BytesLimiter that is made up of child byte limiters?


```
type CompositeBytesLimiter  {
  limiters []BytesLimiter
}

func (c *CompositeBytesLimiter) ReserveWithType(num uint64, dataType storeDataType)   
  for _, l := range c.limiters {
    // ...
  }
}
```

File: pkg/store/bucket.go [hunk 2]
```diff
@@ -3325,8 +3334,7 @@ func (r *bucketIndexReader) LoadSeriesForTime(ref storage.SeriesRef, lset *[]sym
 		return false, errors.Errorf("series %d not found", ref)
 	}
 
-	r.stats.seriesTouched++
-	r.stats.SeriesTouchedSizeSum += units.Base2Bytes(len(b))
+	r.stats.add(SeriesTouched, 1, len(b))
```
Comment (harry671003) [inline pkg/store/bucket.go:3337]: Should the stats refactoring be done in another PR?

File: pkg/store/bucket.go [hunk 3]
```diff
@@ -62,6 +62,8 @@ import (
 	"github.com/thanos-io/thanos/pkg/tracing"
 )
 
+type storeDataType string
```
Comment (harry671003) [inline pkg/store/bucket.go:65]: Assuming that the implementations of the limiter can be done in a separate package (eg: Cortex), shouldn't this type be exported outside this package?

File: pkg/store/bucket.go (hunk 1 above)
Comment (justinjung04) [reply pkg/store/bucket.go:3546]: Yup, that's definitely an option, but I wasn't sure if that extra layer is necessary. I can write it if more ppl are interested.

File: pkg/store/bucket.go (hunk 3 above)
Comment (justinjung04) [reply pkg/store/bucket.go:65]: yup, will update it

File: pkg/store/bucket.go (hunk 2 above)
Comment (justinjung04) [reply pkg/store/bucket.go:3337]: hmm I thought it was small enough to be part of this PR

File: pkg/store/bucket.go [hunk 4]
```diff
@@ -2869,12 +2878,13 @@ func (r *bucketIndexReader) fetchExpandedPostingsFromCache(ctx context.Context,
 	if !hit {
 		return false, nil, nil
 	}
-	if err := bytesLimiter.Reserve(uint64(len(dataFromCache))); err != nil {
-		return false, nil, httpgrpc.Errorf(int(codes.ResourceExhausted), "exceeded bytes limit while loading expanded postings from index cache: %s", err)
+	if bytesLimiter != nil {
```
Comment (yeya24) [inline pkg/store/bucket.go:2881]: Let's remove the nil check here. It is probably redundant after we changing back to use one limiter only

File: pkg/store/limiter.go [hunk 5]
```diff
@@ -93,8 +97,8 @@ func NewSeriesLimiterFactory(limit uint64) SeriesLimiterFactory {
 	}
 }
 
-// NewBytesLimiterFactory makes a new BytesLimiterFactory with a static limit.
-func NewBytesLimiterFactory(limit units.Base2Bytes) BytesLimiterFactory {
+// DefaultBytesLimiterFactory makes a new BytesLimiterFactory with a static limit.
+func DefaultBytesLimiterFactory(limit units.Base2Bytes) BytesLimiterFactory {
```
Comment (yeya24) [inline pkg/store/limiter.go:101]: Is there a reason to rename the factory?

File: pkg/store/bucket.go [hunk 6]
```diff
@@ -62,6 +62,8 @@ import (
 	"github.com/thanos-io/thanos/pkg/tracing"
 )
 
+type StoreDataType string
```
Comment (yeya24) [inline pkg/store/bucket.go:65]: Nit. Is it necessary to use string type here? int is probably fine and we can just do iota.

Comment (justinjung04) [reply pkg/store/bucket.go:65]: true, I originally thought of printing the StoreDataType but not anymore. I'll change it to iota

File: pkg/store/bucket.go (hunk 4 above)
Comment (justinjung04) [reply pkg/store/bucket.go:2881]: Are we confident that bytesLimiter won't be null, as Series/LabelNames/LabelValues functions all create bytes limiter from the factory and pass down? Then I think we should check at least if the factory isn't null, or assign `NewBytesLimiterFactory(0)` if null is passed (and probably similar for Chunks and Series limiter factories). What do you think?

Comment (yeya24) [reply pkg/store/bucket.go:2883]: Since Thanos is only used itself and by downstream projects like Cortex, those project should be responsible to make sure to not return a nil limiter. At least return a noop limiter. Maybe we should add a comment about the contract.

> Then I think we should check at least if the factory isn't null, or assign NewBytesLimiterFactory(0) if null is passed (and probably similar for Chunks and Series limiter factories). What do you think?

But it can be done at the caller side when initializing the bucket store? At least I don't see why we need to do it in Thanos itself since Thanos ensures that the factory is not nil https://github.com/thanos-io/thanos/blob/main/cmd/thanos/store.go#L436.

If it is possible to have nil factory in Cortex we can do it on Cortex. But for me, it should be safe to skip.

Comment (justinjung04) [reply pkg/store/bucket.go]: ok, sounds good to me

File: CHANGELOG.md [hunk 7]
```diff
@@ -27,6 +27,7 @@ We use *breaking :warning:* to mark changes that are not backward compatible (re
 
 - [#7334](https://github.com/thanos-io/thanos/pull/7334) Compactor: do not vertically compact downsampled blocks. Such cases are now marked with `no-compact-mark.json`. Fixes panic `panic: unexpected seriesToChunkEncoder lack of iterations`.
 - [#7393](https://github.com/thanos-io/thanos/pull/7393) *: *breaking :warning:* Using native histograms for grpc middleware metrics. Metrics `grpc_client_handling_seconds` and `grpc_server_handling_seconds` will now be native histograms, if you have enabled native histogram scraping you will need to update your PromQL expressions to use the new metric names.
+- [#7414](https://github.com/thanos-io/thanos/pull/7414) Store: Enhanced bytes limiter with data type param
```
Comment (yeya24) [inline CHANGELOG.md:30]: Let's remove the changelog as there is no user facing changes. Only an internal change for the interface

**Review by yeya24 — APPROVED** [review]
LGTM. Just a small nit on the changelog

**Review by harry671003 — APPROVED** [review]
Great work Justin!

**Comment by yeya24** [comment]
I will merge this change tomorrow if there is no objection or new review comments. Since this change is not user facing at all, we can address in future PRs if you have any other comments

**Review by saswatamcode — APPROVED** [review]
Thanks, generally LGTM! I think it can be put in changelog as your Store error will now tell you the type of data that exceeded bytes limit?

**Comment by justinjung04** [comment]
@saswatamcode 

Thanks for taking a look!

> I think it can be put in changelog as your Store error will now tell you the type of data that exceeded bytes limit?

I haven't changed the error message returned to users, as they already contained the data type information. The main goal of this change was to enhance the limiter logic with access to the data type, which can be hidden from the users.

> Also any plans to expose these to user somehow?

Yes! We have been noticing that resource consumption (CPU and memory) for postings, series and chunks are quite different. For example, we were able to fetch up to 1GB/s of chunks until the CPU is exhausted, while we were only able to fetch 258MB/s of series. We should definitely update the limiter logic to assign different weights per data type.
